src/gui/plugins/environment_visualization/EnvironmentVisualization.cc

```diff
@@ -0,0 +1,312 @@
+  public: bool first {true};
+
+  public: std::atomic<bool> resample{true};
+
+  /////////////////////////////////////////////////
+  public: void CreatePointCloudTopics(
+    std::shared_ptr<components::EnvironmentalData> data) {
+    this->pubs.clear();
+    this->sessions.clear();
+    for (auto key : data->frame.Keys())
+    {
+      this->pubs.emplace(key, node.Advertise<gz::msgs::Float_V>(key));
+      gz::msgs::Float_V msg;
+      this->floatFields.emplace(key, msg);
+      this->sessions.emplace(key, data->frame[key].CreateSession());
+    }
+  }
+
+  /////////////////////////////////////////////////
+  public: void Step(
+    const UpdateInfo &_info,
+    std::shared_ptr<components::EnvironmentalData> &data,
+    const EntityComponentManager& _ecm,
+    double xSamples, double ySamples, double zSamples)
+  {
+    auto now = std::chrono::steady_clock::now();
+    std::chrono::duration<double> dt(now - this->lastTick);
+
+    if (this->resample)
+    {
+      this->CreatePointCloudTopics(data);
+      this->ResizeCloud(data, _ecm, xSamples, ySamples, zSamples);
+      this->resample = false;
+      this->lastTick = now;
+    }
+
+    for (auto &it : this->sessions)
+    {
+      auto res =
+        data->frame[it.first].StepTo(it.second,
+          std::chrono::duration<double>(_info.simTime).count());
+      if (res.has_value())
+      {
+        it.second = res.value();
+      }
+    }
+
+    // Publish at 2 hz for now. In future make reconfigureable.
+    if (dt.count() > 0.5)
+    {
+      this->Visualize(data, xSamples, ySamples, zSamples);
+      this->Publish();
+      lastTick = now;
+    }
+  }
+
+  /////////////////////////////////////////////////
+  public: void Visualize(
+    std::shared_ptr<components::EnvironmentalData> data,
+    double xSamples, double ySamples, double zSamples) {
+
+    for (auto key : data->frame.Keys())
+    {
+      const auto session = this->sessions[key];
+      auto frame = data->frame[key];
+      auto [lower_bound, upper_bound] =
+        frame.Bounds(session);
+      auto step = upper_bound - lower_bound;
+      auto dx = step.X() / xSamples;
+      auto dy = step.Y() / ySamples;
+      auto dz = step.Z() / zSamples;
+      std::size_t idx = 0;
+      for (std::size_t x_steps = 0; x_steps < ceil(xSamples); x_steps++)
+      {
+        auto x = lower_bound.X() + x_steps * dx;
+        for (std::size_t y_steps = 0; y_steps < ceil(ySamples); y_steps++)
+        {
+          auto y = lower_bound.Y() + y_steps * dy;
+          for (std::size_t z_steps = 0; z_steps < ceil(zSamples); z_steps++)
+          {
+            auto z = lower_bound.Z() + z_steps * dz;
+            auto res = frame.LookUp(
+              session, math::Vector3d(x, y, z));
+
+            if (res.has_value())
+            {
+              this->floatFields[key].mutable_data()->Set(idx,
+                static_cast<float>(res.value()));
+            }
+            idx++;
```

Comment:
@arjo129 hmm, I think this code should be setting float field values to NaN if not available in the data frame. The first time around and after a resampling it will behave like that but not if the lookup succeeds at time `t` and fails at time `t + 1`.